### ISSUE TITLE
API improvements?

### ISSUE THREAD

**Issue by wmertens** [gluxon/node-mjpeg-server#7]
I know this implements a `Writable`, but I was mistakenly using it as `fs.createReadStream("img.jpg").pipe(mjpegStream, {end: false})`, which doesn't work, since it needs to send the full size of the file in `_write`.

So now I wonder if one can send an image without specifying size? Or does the mjpeg stream require sizes?

If so, it may be better to expose `.sendFile(path)` and `.sendBuffer(buffer)` and keep the Writable hidden. Here's my `.sendFile`:

```js
MjpegServer.prototype.sendFile = function(path, cb) {
	const onRead = (err, data) => {
		if (err) {return cb && cb(err)}
		this.write(data)
		cb && cb()
	}
	try {
		fs.readFile(path, onRead)
	} catch(err) {
		cb && cb(err)
	}
};
```

Maybe this can be improved, for example by checking if the previous frame was already sent through `res`, and if not, skip the frame? Not sure if that is even possible.

**Comment by gluxon** [comment]
If this issue doesn't have an immediate need to be resolved, let me circle back to this in May. I'd like some time to see if there's any way to implement your suggestions. I'm currently a student and have final exams coming up. Thanks!

**Comment by wmertens** [comment]
No urgency, I copied your code and adjusted it :)

Turns out the length is not needed, but of course that means the marker (`--something\r\n`) can not appear in the JPEG file sent.

I also discovered that Chrome and Safari are waiting for the marker before showing the last sent frame. I fixed that by immediately sending the marker after the file.

So my current version is this:

```js
class MjpegServer {
	constructor(res) {
		this.res = res

		res.writeHead(200, {
			'Content-Type': 'multipart/x-mixed-replace; boundary=MjPeGsTrEaM',
			'Cache-Control': 'no-cache',
			'Connection': 'close',
			'Pragma': 'no-cache'
		})
		this.sendBoundary()
	}
	sendBoundary() {
		this.res.write("--MjPeGsTrEaM\r\nContent-Type: image/jpeg\r\n\r\n");
	}
	writeFile(path, cb) {
		if (this.busy) {return}
		try {
			this.busy = true
			const s = fs.createReadStream(path)
			s.pipe(this.res, {end: false})
			// TODO verify that this also gets called on errors
			s.on('end', () => {
				this.busy = false
				this.sendBoundary()
			})
		} catch(err) {
			cb && cb(err)
		}
	}
	close() {
		this.res.end()
		delete this.res
	}
}
```

I use it like
```js
			const mjpeg = new MjpegServer(res)
			mjpeg.writeFile(cam.lastFrame)
			const onSnapped = ({filename}) => {
				mjpeg.writeFile(filename)
			}
			cam.on('snapped', onSnapped)
			res.once('close', () => {
				cam.removeListener('snapped', onSnapped)
			})
```

By piping a stream, the JPG doesn't need to be loaded into memory first, so it uses less memory. I also (maybe) added frame skipping by not sending files while the previous one is sending. Not sure if that is enough.

I looked into the Node APIs and it seems that the only place to know if a socket is done sending data is by looking at https://nodejs.org/api/net.html#net_socket_buffersize, however, I have no idea if that is already happening. It would be good to somehow test sending a couple of MB of frames to a slow receiver and seeing if the current mechanism is enough to skip frames…

**Comment by wmertens** [comment]
I just tested the throttling locally by setting my Chrome to 2G speed, and it buffered up about 3 frames before skipping frames. So it works fine, no socket magic needed :)

**Comment by gluxon** [comment]
@wmertens We're at a year later, but I would happily accept your changes as a merge request. If you don't have the time, I don't mind using your findings to do it myself. Thanks for your research into this regardless!

**Comment by wmertens** [comment]
@gluxon yes feel free to use it, the code above is still the full version as used in production. No time for PR I'm afraid :-/
